### PR TITLE
continue to flesh out API

### PR DESCRIPTION
**Pull request by oscarbatori**
implement transaction semantics for context manager, add task ID to integration, add Dolt->Metaflow by updating commit message used to store tables

### CODE REVIEW

File: dolt-demos/movie_sentiment_reviews.py [hunk 1]
```diff
@@ -102,16 +102,15 @@ def stats(self):
...
-            with DoltDT(run=self, doltdb_path=self.doltdb_path, branch='vinai/add-rotten-data') as dolt:
-                predictions = pd.Series(predictions).rename('predictions')
-                result = pd.concat([reviews, labels, predictions], axis=1)
-                dolt.write_table(table_name=table_name, df=result, pks=['review'])
-                #dolt.commit_table_writes()
-        
-        write_prediction_to_table(self.train_reviews, self.train_labels, train_predictions, "train_results")
-        write_prediction_to_table(self.test_reviews, self.test_labels, test_predictions, "test_results")
-        
+        def write_prediction_to_table(dolt_context: DoltDT, reviews, labels, predictions, table_name):
```
Comment (VinaiRachakonda) [inline dolt-demos/movie_sentiment_reviews.py:105]: Are we officially calling the DoltDT a Dolt Context now? If so, I think now is the time to take a step back and write a design doc to really hone down on our thinking on what we want to model and what terms we want to use. 

Not sure I like all the new terminology that we're introducing.

File: metaflow/datatools/dolt.py [hunk 2]
```diff
@@ -64,16 +66,31 @@ def __init__(self, run, doltdb_path: str, branch: str = 'master'):
 
     def __enter__(self):
         assert isinstance(self.run, FlowSpec) and current.is_running_flow, 'Context manager use requires running flow'
+        assert self.doltdb.status().is_clean, 'DoltDT as context manager requires clean working set for transaction semantics'
```
Comment (VinaiRachakonda) [inline metaflow/datatools/dolt.py:69]: Not sure what this sentence means.

File: metaflow/datatools/dolt.py [hunk 3]
```diff
@@ -160,10 +184,10 @@ def _get_tables_for_access_records(self,
                                        steps: List[str]) -> Mapping[str, Mapping[str, pd.DataFrame]]:
         result = {}
         for access_record in access_records:
-            if runs and access_record.run_id not in runs and steps and not access_record.step not in steps:
+            if runs and access_record.run_id not in runs and steps and not access_record.step_name not in steps:
```
Comment (VinaiRachakonda) [inline metaflow/datatools/dolt.py:188]: can you leave a comment explaining this condition?

File: metaflow/runtime.py [hunk 4]
```diff
@@ -271,7 +271,7 @@ def _queue_task_join(self, task, next_steps):
         if len(next_steps) > 1:
             msg = 'Step *{step}* transitions to a join and another '\
                   'step. The join must be the only transition.'
-            raise MetaflowInternalError(task, msg.format(step=task.step))
+            raise MetaflowInternalError(task, msg.format(step=task.step_name))
```
Comment (VinaiRachakonda) [inline metaflow/runtime.py]: Are we sure we aren't affecting other parts of codebase with this change? What is the intention here?

**Review by VinaiRachakonda — CHANGES_REQUESTED** [review]

File: metaflow/runtime.py (hunk 4 above)
Comment (oscarbatori) [reply metaflow/runtime.py]: Good spot, this is a PyCharm snafu, will remove.

**Comment by oscarbatori** [comment]
So, I think you are right the terminology is a bit unclear here, but I don't agree that we have come upon anything incoherent.

There are two ways to use the `DoltDT` class:
1. context manager inside a flow, via `with DoltDT(...) as dolt`
2. directly instantiating it via `DoltDT(...)` as a way to query Dolt through the Metaflow metadata management system

The first UX supports users defining flows, the second supports them querying data in Dolt via the flows that read or wrote that data.

Does that make sense? I think a design doc in general is a good idea, but I don't think this is especially confusing or controversial. I also think actually writing unit tests to make the semantics explicit will go along way to resolving the confusion.

File: metaflow/datatools/dolt.py (hunk 2 above)
Comment (oscarbatori) [reply metaflow/datatools/dolt.py:69]: It means that when you instantiate `DoltDT` as a context manager the writes performed by that context manager constitute a transaction: either they all succeed and a commit is created, or they all fail the state of the Dolt database is unchanged.

Each task that has a Dolt context manager has its writes encapsulated in a transaction, essentially.

**Pull request merged**